### ISSUE TITLE
Component API draft

### ISSUE THREAD

**Issue by lbialy** [VirtusLab/besom#85]
I aim to establish a structure where an API would function in the following manner:

```scala
case class Postgres(urn: Output[String], jdbcUrl: Output[String], username: Output[String], password: Output[String]) extends ComponentResource
 
component[Postgres]("name", "com.virtuslab:besom-templates:postgres") { 
  val postgresOperator = deployment("postgres-operator", ...)
  val postgresCRD = crd("postgres-crd", ...)

  val deploymentData = fetchDeploymentFromCrd(postgresCRD)
  Postgres(componentUrn(), deploymentData.jdbcUrl, deploymentData.username, deploymentData.password)
}
```

In the context above:

```scala
  val postgresOperator = deployment("postgres-operator", ...)
  val postgresCRD = crd("postgres-crd", ...)
```

These take `ResourceOptions` and I must enforce that they have a parent set to the component. I'm considering accomplishing this through Context:
```scala
def component[A <: ComponentResource: RegistersOutputs](name: NonEmptyString, typ: ResourceType)(f: Context ?=> Output[A]): Output[A]
```
Here, `RegistersOutputs` only serves to transform the given `A` into `Map[String, Output[?]]` and pass it to `ctx.registerResourceOutputs`.

The crucial part here is the context within the scope of function `f` where the user declares resources being children of this component, it must override `getParent`. Then, the component should automatically propagate.

In other words, the context that we pass to `f` will be a wrapper over the normal context.
